pkg/pool-weighted/contracts/managed/ManagedPoolSettings.sol

```diff
@@ -853,75 +855,58 @@ abstract contract ManagedPoolSettings is BasePool, ProtocolFeeCache, ReentrancyG
      * in some scenarios to account for the fact that the Pool now has fewer tokens. This is a permissioned function.
      * @param token - The ERC20 token to be removed from the Pool.
-     * @param recipient - The address to receive the Pool's balance of `token` after it is removed.
      * @param burnAmount - The amount of BPT to be burned after removing `token` from the Pool.
-     * @param minAmountOut - Will revert if the number of tokens transferred from the Vault is less than this value.
-     * @return The amount of tokens the Pool held, sent to `recipient`.
      */
     function removeToken(
         IERC20 token,
-        address recipient,
-        uint256 burnAmount,
-        uint256 minAmountOut
-    ) external authenticate nonReentrant whenNotPaused returns (uint256) {
-        // We require the pool to be initialized (shown by the total supply being nonzero) in order to remove a token,
-        // maintaining the behaviour that no exits can occur before the pool has been initialized.
-        // This prevents the AUM fee calculation being triggered before the pool contains any assets.
+        address sender,
+        uint256 burnAmount
+    ) external authenticate nonReentrant whenNotPaused {
         _require(totalSupply() > 0, Errors.UNINITIALIZED);
 
         // To reduce the complexity of weight interactions, tokens cannot be removed during or before a weight change.
+        // This is for symmetry with addToken.
         _ensureNoWeightChange();
 
-        // Exit the pool, returning the full balance of the token to the recipient
-        (IERC20[] memory tokens, uint256[] memory unscaledBalances, ) = getVault().getPoolTokens(getPoolId());
+        // Before this function is called, the caller must have withdrawn all balance for `token` from the Pool. This
+        // means that the Pool is in an invalid state, since among other things the invariant is zero. Because we're not
+        // in a valid state and all value-changing operations will revert, we are free to modify the Pool state (e.g.
+        // alter weights).
+        // We don't need to test the zero balance since the Vault will simply revert on deregistration if this is not
+        // the case.
+
+        // Removing a token will cause for the weights of all other tokens to increase. This is fine, as there is no
+        // maximum weight. We also don't need to check that the new token exists in the Pool, as the Vault will simply
+        // revert if we try to deregister a token that is not registered.
+        // We do, however, want to check that the Pool will end up with at least two tokens. This simplifies some
+        // assumptions made elsewhere (e.g. the denormalized weight sum will always be non-zero), and doesn't greatly
+        // restrict the controller.
+
+        (IERC20[] memory tokens, , ) = getVault().getPoolTokens(getPoolId());
```

Comment:
There may be a cheaper way to check the length of this array. scaling factors perhaps?

Reply:
We don't track any length ourselves - the scaling factors are stored in mapping entries. This is because we favor 1 SLOAD access in swaps (where we know the addresses of the tokens involved), at the cost of having to fetch the tokens array in other operations such as joins.

Reply:
Yeah, forgot that _scalingFactors begins with this exact query.